### ISSUE TITLE
DynamoDB takes about 3s for any query on a small table when AWSMobileClient instance is used as credentials provider

### ISSUE THREAD

**Issue by markorakita** [aws-amplify/aws-sdk-android#944]
**State your question**
I have a small table in DynamoDB with only 2 items, each item containing 7 attributes, 2 strings and 5 numbers. I've tried calling scan from DynamoDBMapper, it took ~3s. I've tried query it took the same. Tried load for just one item specifying the hash key, took the same. I am obviously doing something wrong, there is no way DynamoDB is this slow.

- Is sdk for ddb maybe slow from debug mode?
- I have 5 rcu, I believe it is enough for such a small table so it shouldn't affect the speed?
- I am in central Europe, at first my table was in us-east-1 so I transferred it to eu-central-1 now, but the speed is just slightly improved, if any. My app was created via the mobile hub, and everything was setup in us-east-1 (cognito, S3, etc). Maybe it still makes cross region calls somehow (for example if ddb interacts with cognito for auth)?

I have no other ideas, please help :)

**Which AWS Services are you utilizing?**
DynamoDB database, called via DynamoDBMapper from Android AWS SDK.

**Provide code snippets (if applicable)**
This is what I used to scan table:
```
AmazonDynamoDBClient dynamoDBClient = new AmazonDynamoDBClient(AWSMobileClient.getInstance());
DynamoDBMapper  dynamoDBMapper = DynamoDBMapper.builder()
                .dynamoDBClient(dynamoDBClient)
                .awsConfiguration(AWSMobileClient.getInstance().getConfiguration())
                .build();
PaginatedList<ShopItemsDO> models = dynamoDBMapper.scan(ShopItemsDO.class, new DynamoDBScanExpression());
models.loadAllResults();
return new ArrayList<>(models);
```
This is what I used to query table:
```
ShopItemsDO kv = new ShopItemsDO();
kv.setItemName("Chair");
PaginatedList<ShopItemsDO> models = dynamoDBMapper.query(ShopItemsDO.class, new DynamoDBQueryExpression<ShopItemsDO>().withHashKeyValues(kv).withConsistentRead(false));
models.loadAllResults();
return new ArrayList<>(models);
```
This is what I used to load item from a table:
`ShopItemsDO kv = dynamoDBMapper.load(ShopItemsDO.class, "Chair", "Large");`

**Environment(please complete the following information):**
 - SDK Version: 2.13.2

**Device Information (please complete the following information):**
 - Device: Huawei Mate 10pro
 - Android Version: Pie 9.0
 - Specific to simulators: No

**Comment by markorakita** [comment]
I've now tried doing AWSMobileClient initialization first, and then doing database calls, and guess what, DynamoDB calls finish in a miliseconds. This seems like a bug to me.

Here is what I am currently doing:
```
AWSMobileClient.getInstance().initialize(applicationContext, new Callback<UserStateDetails>()
        {
            @Override
            public void onResult(UserStateDetails result)
            {
                MyActivity.this.runOnUiThread(new Runnable()
                {
                    @Override
                    public void run()
                    {
                        queryDynamoDB();
                    }
                });
            }

            @Override
            public void onError(Exception e) {}
        });
```

Function queryDynamoDB() itself now takes ~3s (not including initialization time).

If I do it separate, on one button click I do the initialization, on another button click I do the queryDynamoDB(), then queryDynamoDB() takes only fraction of a second.

Btw queryDynamoDB() launches new async task inside of which I do query of one item from a DynamoDB. I have breakpoints at the task launch, and when task is finished, so I can measure the time.

**Comment by markorakita** [comment]
As I can see in the AWSMobileClient _initialize() function code, only thing that happens after calling callback.onResult() is setUserState():
```
callback.onResult(userStateDetails);
setUserState(userStateDetails);
```

Could it somehow affect DynamoDB client when it is initalized inside callback.onResult() with AWSMobileClient which doesn't have user state set yet?

**Comment by markorakita** [comment]
I've debugged further, and I found out that AWSMobileClient.getCredentials() is being called for ddb functions, and that is what causes the 3s hang. In a scenario where I was doing AWSMobileClient.initialize() separately, I was also doing the AWSMobileClient.signIn(), so that already took care of credentials. That explains why ddb calls take so long in a scenario where they are executed inside AWSMobileClient.initialize() onResult(), when app is cold started.

Is there any way to prevent this? Why does AWSMobileClient.getCredentials() take so long to execute?

**Comment by markorakita** [comment]
Ok, now I've realized why people use CognitoCachingCredentialsProvider in their code.

I've changed this:
`AmazonDynamoDBClient dynamoDBClient = new AmazonDynamoDBClient(AWSMobileClient.getInstance());`

To this:
```
CognitoCachingCredentialsProvider ccp = new CognitoCachingCredentialsProvider(context, AWSMobileClient.getInstance().getConfiguration());
AmazonDynamoDBClient dynamoDBClient = new AmazonDynamoDBClient(ccp);
```

Now everything works fast. Reason is that AWSMobileClient.getCredentials() implementation calls waitForSignIn() which calls getUserStateDetails(false) where **false** means "don't use offline cashed credentials".

Sorry for this long thread, but I wouldn't lose this many hours on this if you had more than 1 example of how AmazonDynamoDBClient for android sdk is initialized. And in that [1 example](https://docs.aws.amazon.com/aws-mobile/latest/developerguide/mobile-hub-add-aws-mobile-nosql-database.html), you initialize it with AWSMobileClient.getInstance() as credentials provider. [Here](https://aws-amplify.github.io/docs/android/start) you don't even mention DynamoDB API.

**More documentation on this SDK really wouldn't hurt anyone.** I had to dive deep into your source code on so many occasions just to understand how to use some basic APIs...

I will close this thread now and open new thread for questions about credentials provider, since I am not sure if using CognitoCachingCredentialsProvider is a way to go based on this comment in AmazonDynamoDBClient constructor:
```
     * If AWS session credentials are passed in, then those credentials will be
     * used to authenticate requests. Otherwise, if AWS long-term credentials
     * are passed in, then session management will be handled automatically by
     * the SDK. Callers are encouraged to use long-term credentials and let the
     * SDK handle starting and renewing sessions.
```
